File: lib/pdfToMarkdown.ts
import { mkdtemp, rm, writeFile } from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'
import { PythonToolError, runPythonTool } from './pythonTools'

/**
 * PDF → structured Markdown, the first tool on the Werkzeuge page (app/tools).
 *
 * The extraction itself happens in Python (python/pdf_to_markdown.py, pymupdf4llm)
 * because nothing in the JS ecosystem reconstructs headings, lists and tables
 * from a PDF's glyph soup nearly as well. This module owns everything around
 * it: the temp file the subprocess reads, the guards, and the translation of
 * the helper's error codes into messages the UI can show as-is.
 */

/** Same cap the other importers use — see python/pdf_to_markdown.py. */
export const MAX_PDF_MARKDOWN_CHARS = 100_000

/** Upload cap. Well past any paper; a scanned book would only yield noise anyway. */
export const MAX_PDF_BYTES = 30 * 1024 * 1024

/** Generous: a 300-page PDF takes a few seconds, a pathological one must still end. */
const CONVERT_TIMEOUT_MS = 180_000

export interface PdfMarkdown {
  markdown: string
  pages: number
  chars: number
  /** True when the text was cut at MAX_PDF_MARKDOWN_CHARS. */
  truncated: boolean
  /** The PDF's metadata title — often empty, often junk ("Microsoft Word - …"). */
  title: string
}

/** The helper's JSON contract. */
type HelperResult =
  | { ok: true; markdown: string; pages: number; chars: number; truncated: boolean; title: string }
  | { ok: false; error: string; detail?: string }

/** User-facing text per handled helper error. */
const ERROR_MESSAGES: Record<string, string> = {
  not_a_pdf:          'Die Datei konnte nicht als PDF gelesen werden.',
  encrypted:          'Das PDF ist passwortgeschützt und lässt sich nicht auslesen.',
  conversion_failed:  'Das PDF konnte nicht umgewandelt werden.',
}

/** A PDF always starts with `%PDF-`; catch a mislabelled file before spawning. */
function looksLikePdf(bytes: Buffer): boolean {
  return bytes.subarray(0, 5).toString('latin1') === '%PDF-'
}

/**
 * Converts an uploaded PDF. Throws `PythonToolError` with a German message that
 * is safe to show the user; the `detail` stays for the server log.
 */
export async function pdfToMarkdown(bytes: Buffer): Promise<PdfMarkdown> {
  if (bytes.length === 0)         throw new PythonToolError('Die Datei ist leer.')
  if (bytes.length > MAX_PDF_BYTES) {
    throw new PythonToolError(
      `Die Datei ist größer als ${Math.round(MAX_PDF_BYTES / 1024 / 1024)} MB.`,
    )
  }
  if (!looksLikePdf(bytes)) throw new PythonToolError('Die Datei ist kein PDF.')

  // Own directory per run: the subprocess only ever sees this one file, and the
  // cleanup is a single recursive remove that cannot miss a stray sibling.
  const dir = await mkdtemp(path.join(os.tmpdir(), 'nuggets-pdf-'))
  const file = path.join(dir, 'upload.pdf')
  try {
    await writeFile(file, bytes)
    const result = await runPythonTool<HelperResult>(
      'pdf_to_markdown.py',
      [file, '--max-chars', String(MAX_PDF_MARKDOWN_CHARS)],
      { timeoutMs: CONVERT_TIMEOUT_MS },
    )

    if (!result.ok) {
      throw new PythonToolError(
        ERROR_MESSAGES[result.error] ?? 'Das PDF konnte nicht umgewandelt werden.',
        result.detail,
      )
    }

    return {
      markdown:  result.markdown,
      pages:     result.pages,
      chars:     result.chars,
      truncated: result.truncated,
      title:     result.title,
    }
  } finally {
    await rm(dir, { recursive: true, force: true }).catch(() => {})
  }
}

File: python/pdf_to_markdown.py
#!/usr/bin/env python3
"""Extract a PDF's text as structured Markdown (headings, lists, tables).

Invoked as a short-lived subprocess by lib/pdfToMarkdown.ts — never long-running,
never listening on a port. The contract with the caller is deliberately narrow:

  * argv:   <pdf-path> [--max-chars N]
  * stdout: exactly ONE JSON object, nothing else
  * stderr: free-form (MuPDF warns there); only ever surfaced in server logs
  * exit:   0 = handled (read `ok`), anything else = crashed, stdout unusable

Errors the user can act on (encrypted file, not a PDF) come back as
`{"ok": false, ...}` with exit 0, so the route can show a real message instead
of a generic 502.
"""

import argparse
import json
import re
import sys

# Same cap the other importers use (/api/extract, lib/googleDrive.ts): a nugget
# is a note, not a book, and the AI revision downstream pays per character.
DEFAULT_MAX_CHARS = 100_000

_LIST_ITEM = re.compile(r"^\s*(?:[-*+]|\d+[.)])\s")


def clean_markdown(md: str) -> str:
    """Tidy the raw conversion into something a human wants to read.

    pymupdf4llm is faithful to the PAGE, which on a text level shows up as
    artefacts a reader never saw: runs of blank lines where a column ended,
    trailing spaces from justified text, and — because headings are set in a
    bold font — every heading additionally wrapped in `**`. Structure itself
    (heading levels, lists, tables) is never touched; that is the part we asked
    pymupdf4llm for.
    """
    md = md.replace("\r\n", "\n").replace("\r", "\n")
    # Trailing whitespace, but keep Markdown's two-space hard line break intact.
    md = re.sub(r"[ \t]+$", "", md, flags=re.MULTILINE)
    # `## **Titel**` → `## Titel` (only when the bold spans the whole heading —
    # a heading with one bolded word inside keeps it).
    md = re.sub(r"^(#{1,6} )\*\*(.+?)\*\*$", r"\1\2", md, flags=re.MULTILINE)
    # Collapse 3+ blank lines to one blank line.
    md = re.sub(r"\n{3,}", "\n\n", md)
    return tighten_lists(md.strip())


def tighten_lists(md: str) -> str:
    """Drop blank lines BETWEEN consecutive list items.

    pymupdf4llm separates every bullet by a blank line, which Markdown reads as
    a "loose" list and renders with a `<p>` inside each `<li>` — visibly airier
    than the same list typed by hand. Only blank lines with a list item on both
    sides go; a blank line before or after the list is structural and stays.
    """
    lines = md.split("\n")
    out: list[str] = []
    for i, line in enumerate(lines):
        if (
            line.strip() == ""
            and out
            and _LIST_ITEM.match(out[-1])
            and i + 1 < len(lines)
            and _LIST_ITEM.match(lines[i + 1])
        ):
            continue
        out.append(line)
    return "\n".join(out)


def main() -> int:
    parser = argparse.ArgumentParser(description="PDF → structured Markdown")
    parser.add_argument("path", help="path to the PDF file")
    parser.add_argument("--max-chars", type=int, default=DEFAULT_MAX_CHARS)
    args = parser.parse_args()

    # Imported late so an --help / argv error does not pay the ~1 s import cost.
    import pymupdf
    import pymupdf4llm

    try:
        doc = pymupdf.open(args.path)
    except Exception as exc:  # noqa: BLE001 — any open failure means "not a usable PDF"
        print(json.dumps({"ok": False, "error": "not_a_pdf", "detail": str(exc)}))
        return 0

    try:
        # A password-protected file opens fine but yields empty pages, which
        # would look like "PDF without text" — name the real cause instead.
        if doc.needs_pass:
            print(json.dumps({"ok": False, "error": "encrypted"}))
            return 0

        page_count = doc.page_count
        title = (doc.metadata or {}).get("title") or ""

        markdown = pymupdf4llm.to_markdown(
            doc,
            # Nugget content is deliberately image-free (PLAN.md TODO 5), and
            # embedded images would arrive as base64 data: URIs — the exact
            # ballast cleanExportedMarkdown strips on the Google Docs path.
            ignore_images=True,
            write_images=False,
            embed_images=False,
            # Graphics stay ON: table borders ARE vector graphics, and dropping
            # them costs the table detection that makes this better than a
            # plain text dump.
            table_strategy="lines_strict",
            show_progress=False,
        )
    except Exception as exc:  # noqa: BLE001 — report, don't traceback into stdout
        print(json.dumps({"ok": False, "error": "conversion_failed", "detail": str(exc)}))
        return 0
    finally:
        doc.close()

    markdown = clean_markdown(markdown)
    truncated = len(markdown) > args.max_chars
    if truncated:
        markdown = markdown[: args.max_chars]

    print(json.dumps({
        "ok": True,
        "markdown": markdown,
        "pages": page_count,
        "chars": len(markdown),
        "truncated": truncated,
        "title": title.strip(),
    }))
    return 0


if __name__ == "__main__":
    sys.exit(main())
